clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java

```diff
@@ -690,6 +690,20 @@ public Node provide() {
             this(false, callName, deadlineMs, nodeProvider);
         }
 
+        Call(String callName, long deadlineMs, NodeProvider nodeProvider,
```

Comment:
One of the annoying aspects of the code below is that we have a lot of redundant logic for constructing a call in the context of a previously failed call. I wonder if it would simplify the logic if we added a constructor which accepts the failed call as an argument and then adjusts `tries` etc accordingly.

Reply:
Yeah that was something that was bothering me as well. However when I tried to create a constructor that would take the failed call, it wasn't working as to use the constructor I also had to implement the methods in the `Call` class. That's why I think we get the call normally from the method that is returning the call. Taking the `getDescribeConsumerGroupsCall` example, it could get a bit more messy if wanted to implement the methods within the new constructor:

```java
                if (ConsumerGroupOperationContext.hasCoordinatorMoved(response)) {
                    Call call = this.call();
                    Call call2 = new Call(call) {
                        @Override
                        Builder createRequest(final int timeoutMs) {
                            return call.createRequest(timeoutMs);
                        }

                        @Override
                        void handleResponse(final AbstractResponse abstractResponse) {
                            call.handleResponse(abstractResponse);
                        }

                        @Override
                        void handleFailure(final Throwable throwable) {
                            call.handleFailure(throwable);
                        }
                    };
                    rescheduleFindCoordinatorTask(context, () -> call2);
                    return;
                }
```

Thinking about this a bit more though, another way I was thinking we could simplify the logic without having to reimplement the methods for the constructor was like this (taking the same example of the `getDescribeConsumerGroupsCall`:

```java
// in the handleResponse section
if (ConsumerGroupOperationContext.hasCoordinatorMoved(response)) {
     Call call = getDescribeConsumerGroupsCall(context);
     call.updateTriesAndNextAllowedTryMs();
     rescheduleFindCoordinatorTask(context, () -> call2);
     return;
}
```

and the `Call` class would get this new method:

```java
void updateTriesAndNextAllowedTryMs() {
            this.tries += 1;
            this.nextAllowedTryMs = calculateNextAllowedRetryMs();
}
```

Personally, I prefer to keep the code more readable by getting the call from the method and simply using the internal method to update it - however if you prefer we use the constructor way, I can change it to that.

Reply:
Hmm.. I see. Yeah, it seems a bit awkward. Let me offer one more suggestion before I give up. I think the awkwardness is a result of the fact that we cannot easily chain the retry context from one call to another. Suppose that we create a separate object for this:
```java
class ApiContext {
  long deadlineMs;
  int tries;
  long nextAllowedTryMs;

  ApiContext retry(long now) {
    return new ApiContext(
      deadlineMs,
      tries + 1,
      now + backoffMs);
  }
}
```
Would that work?

Reply:
That could work, but is there a reason we don't want to directly modify the properties in the call upon retry by having a method inside `Call` that does that? Something I was thinking about with   having a `CallContext` across calls is that would it be restricted to these individual properties? Or do we expect that the properties that might have to be chained across calls could increase in the future?

Reply:
Mainly just trying to consolidate the logic for updating the retry state so that we don't have to remember to do it for new APIs. Also, eventually we want to have exponential backoff logic. I was thinking `CallContext` would be a good way to encapsulate any additional state that we would need.

Reply:
Yeah that's a fair point, the only thing is that those properties are directly accessed in a lot of places (when it comes to failure and retry logic) and I'm thinking that changing the way those properties are encapsulated would require refactoring that could possibly be beyond the scope of this PR. 

For example, in the `fail` method of the `Call` class, these attributes are directly accessed many times to determine whether or not to actually fail the call or retry it. Also, I'm considering how changing the encapsulation would affect all other calls outside of the consumer groups operations, as those would also have to be updated to use the `CallContext` object instead.

The other option I'm thinking (to reduce the amount of refactoring) would be to have both `CallContext` and the properties at play, but that would be a lot of redundant information and I'm not quite sure if we'd want that. 

With either case, I can change it to be either way, but I personally think having a context for the `Call` is something that should be done so we can more easily chain it across calls, but perhaps is out of scope for this PR and could be picked up in a future PR.

Reply:
Ok, let's leave this as a potential future improvement (perhaps as part of the the exponential backoff kip).